text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+
+    unsafe fn grow_zeroed(
+        &self,
+        _handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        debug_assert!(
+            new_layout.size() >= old_layout.size(),
+            "{new_layout:?} must have a greater size than {old_layout:?}"
+        );
+
+        Self::validate_layout(new_layout)?;
+
+        let pointer = self.0.get() as *mut u8;
+
+        //  Safety:
+        //  -   Both starting and resulting pointers are in bounds of the same allocated objects as `old_layout` fits
+        //      `pointer`, as per the pre-conditions of `grow_zeroed`.
+        //  -   The offset does not overflow `isize` as `old_layout.size()` does not.
+        let pointer = unsafe { pointer.add(old_layout.size()) };
+
+        //  Safety:
+        //  -   `pointer` is valid, since `self` is valid.
+        //  -   `pointer` points to at an area of at least `mem::size_of::<T>() - old_layout.size()`.
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StableStore for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
```

Comment:
That's a downside I had not considered.

The benefits I see in not requiring `Send`/Sync` in the handle precisely require that the user NOT ignore those properties of the handle, which would make `NonNull<u8>` unusable as a handle and require wrapping.

Reply:
The thing is, `NonNull<T>` isn't "really" not `Send`/`Sync`.

One way or the other, someone loses. In this case where consumers are supposed to know to not bound their own send/sync on the handles', using actually not send/sync types like `Cell` as a handle is a *giant* soundness footgun.

Respecting the not send/sync of handles and using `NonNull` as a handle results in `Box<T, Global>` not being send/sync, because it contains `NonNull`.

Providing a `Send`/`Sync` `struct AllocHandle(NonNull<u8>);` isn't a large burden.

Reply:
In that case the trait would just use `AllocHandle` instead of `NonNull<u8>`.